sherlock/internal/models/role_auto_assign.go

```diff
@@ -0,0 +1,72 @@
+package models
+
+import (
+	"context"
+	"fmt"
+	"github.com/broadinstitute/sherlock/go-shared/pkg/utils"
+	"github.com/broadinstitute/sherlock/sherlock/internal/clients/slack"
+	"gorm.io/gorm"
+	"time"
+)
+
+// KeepAutoAssigningRoles will periodically add Users to Roles that are marked with
+// Role.AutoAssignAllUsers who do not already have a RoleAssignment. See
+// Role.AutoAssignAllUsers.
+//
+// This function isn't a perfect fit in the models package, but it's not a good fit
+// for role_propagation or suitabiltiy_synchronization either. Rather than making
+// a new package for this one function, we just file it under "model management"
+// and put it here.
+func KeepAutoAssigningRoles(ctx context.Context, db *gorm.DB) {
+	for {
+		select {
+		case <-ctx.Done():
```

Comment:
Exactly -- it's entirely possible, but that's possible across role propagation, suitability synchronization, metrics calculation, etc. The design here is the same as there: each pass is independent so it'll get caught by the next one. This helps with the whole class of issues with transient errors or a database restart or whatever.

Reply:
I did just push a commit to add "select case <-ctx.Done: return" to the inner loops too, which should help this function bail out quicker during a shutdown. It would be okay for this to end up getting killed by the shutdown sequence (as in, what I just added isn't strictly necessary) but it might as well be as graceful as possible.